src/lua/core/Import.cpp

```diff
@@ -48,17 +95,267 @@ static std::string get_caller(lua_State *L)
+	// e.g. '/data/modules/path/to/file.lua' would become 'data.modules.path.to'
+	size_t start = 0;
+	size_t end = path.find_first_of('/', 1); // use start=1 to generate string::npos when the path == "/"
+
+	while (end != std::string::npos) {
+		auto fragment = path.substr(start, end - start);
+
+		if (fragment.size() > 0) {
+			// Paths containing the separator character cannot be translated into a module name
+			// this means you can't do implicit lookup using a path starting in a hidden directory, e.g.
+			// /data/.local/mylib.lua
+			if (fragment.find('.') != std::string::npos)
+				return "";
+			module_name += module_name.empty() ? fragment : '.' + fragment;
+		}
+
+		start = end + 1;
+		end = path.find_first_of('/', start);
+	}
+
+	return module_name;
+}
+
+// If the given file path is cached, returns the cached object on the stack.
+// Otherwise if the path points to a valid file, executes that file and sets
+// the result in the module cache.
+// In all cases, if this function returns false the stack is unmodified.
+static bool load_file_cached(lua_State *L, const std::string &path, int cacheIdx)
+{
+	LUA_DEBUG_START(L);
+	cacheIdx = lua_absindex(L, cacheIdx);
+
+	// If the object is in the import cache, return it.
+	lua_getfield(L, cacheIdx, path.c_str());
+	if (!lua_isnil(L, -1)) {
+		DEBUG_INDENTED_PRINTF("-> loaded module file %s from cache.\n", path.c_str());
+		LUA_DEBUG_END(L, 1);
+		return true;
+	} else
+		lua_pop(L, 1);
+
+	// FileSystem::JoinPathBelow can't join a path starting with '/' for some reason
+	// TODO: find out why and remove that restriction.
+	std::string filePath = (path[0] == '/') ? path.substr(1) : path;
+	auto fileInfo = FileSystem::gameDataFiles.Lookup(filePath);
+	if (fileInfo.Exists()) {
+		// Load and run the file from disk
+		auto data = fileInfo.Read();
+		pi_lua_dofile(L, *data, 1);
+
+		DEBUG_INDENTED_PRINTF("-> loaded module file %s from disk\n", path.c_str());
+		DEBUG_INDENTED_PRINTF("-> module returned a %s value\n", lua_typename(L, lua_type(L, -1)));
+
+		// Since we loaded the module, we don't want to re-load it again.
+		// Replace the nil return value with 'true' to signify a successful
+		// module load
+		if (lua_isnil(L, -1)) {
+			lua_pop(L, 1);
+			lua_pushboolean(L, true);
+		}
+
+		// Store the module in the import cache
+		lua_pushvalue(L, -1);
+		lua_setfield(L, cacheIdx, path.c_str());
+
+		LUA_DEBUG_END(L, 1);
+		return true;
+	}
+
+	LUA_DEBUG_END(L, 0);
+	return false;
+}
+
+// Check the path cache for a module name to filesystem path mapping
+// If present, this module has been successfully loaded before.
+static std::string get_path_cache(lua_State *L, const std::string &moduleName)
+{
+	LUA_DEBUG_START(L);
+
+	lua_getfield(L, LUA_REGISTRYINDEX, MODULE_MAP_CACHE_NAME);
+	lua_getfield(L, -1, moduleName.c_str());
+	std::string path;
+	if (lua_isstring(L, -1))
+		path = lua_tostring(L, -1);
+	lua_pop(L, 2);
+	LUA_DEBUG_END(L, 0);
+	return path;
+}
```

Comment:
Dang, when I read code like this I'm glad to be mainly programming in Rust and C these days. I'm guessing the cache miss signal is the empty string ?

Reply:
Yeah, I consider it semantically valid to signal the lack of a cached path for the module by simply returning an empty string.